File: Solutions/865-Smallest Subtree with all the Deepest Nodes/865-Smallest Subtree with all the Deepest Nodes.cs
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int val=0, TreeNode left=null, TreeNode right=null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
public class Solution {
    public TreeNode SubtreeWithAllDeepest(TreeNode root) {
        int d = GetDepth(root.left) - GetDepth(root.right);
        if(d == 0)
            return root;
        else if(d > 0)
            return SubtreeWithAllDeepest(root.left);
        else 
            return SubtreeWithAllDeepest(root.right);
    }
    
    private int GetDepth(TreeNode root){
        if(root == null)
            return 0;
        int left = GetDepth(root.left);
        int right = GetDepth(root.right);
        return 1 + Math.Max(left, right);
    }
}

//Fourth fastest
public class Solution {
       public class SubTree
        {
            public TreeNode parent { get; set; }
            public int level { get; set; }
            public SubTree(TreeNode p,int l)
            {
                parent = p;
                level = l;
            }
        }

        public SubTree FindSubTree(TreeNode root,int level, TreeNode parent)
        {
            if (root == null)
                return new SubTree(root, level);
            SubTree left = FindSubTree(root.left, level + 1, root);
            SubTree right = FindSubTree(root.right, level + 1, root);
            if (left.level > right.level)
            {
                return left;
            }
            else if (left.level < right.level)
            {
                return right;
            }
            return new SubTree(root, left.level);
        }

        public TreeNode SubtreeWithAllDeepest(TreeNode root)
        {
            SubTree subtree = FindSubTree(root,0, root);
            if (subtree != null)
                return subtree.parent;
            return null;
        }
}

//Third fastest
public class Solution {
    public TreeNode SubtreeWithAllDeepest(TreeNode root) {
        var maxDepth = FindDepth(root);
        return FindSubtree(root, 1, maxDepth);
    }
    
    TreeNode FindSubtree(TreeNode node, int depth, int maxDepth){
        if(null == node) return null;
        
        var isLeaf = (null == node.left && null == node.right);
        if(isLeaf && depth == maxDepth)
            return node;
        
        var left = FindSubtree(node.left, depth+1, maxDepth);
        var right = FindSubtree(node.right, depth+1, maxDepth);
        if(null != left && null != right)
            return node;
        return left ?? right;
    }
    
    int FindDepth(TreeNode node){
        if(null == node) return 0;
        return 1 + Math.Max(FindDepth(node.left), FindDepth(node.right));
    }
}

//Second fastest
public class Solution {
    public TreeNode SubtreeWithAllDeepest(TreeNode root) {
        var (_, node) = SubtreeWithAllDeepest(root, 0);
        
        return node;
    }
    
    public (int,TreeNode) SubtreeWithAllDeepest(TreeNode root, int depth) {
        if(root == null)
        {
            return (-1, null);
        }
        
        var (lDepth, lNode) = SubtreeWithAllDeepest(root.left, depth+1);
        var (rDepth, rNode) = SubtreeWithAllDeepest(root.right, depth+1);
        
        if(lDepth == rDepth)
        {
            if(lDepth == -1)
            {
                return (depth, root);
            }
            else
            {
                return (lDepth, root);
            }
        }
        
        if(lDepth>rDepth)
            return (lDepth, lNode);
        else
            return (rDepth, rNode);
    }
}

//Fastest
public class Solution {
    public TreeNode SubtreeWithAllDeepest(TreeNode root) {
        var result = dfs(root); 
        return result.Item1;
    }
    
    private Tuple<TreeNode, int>dfs(TreeNode node) {
        if (node == null) {
            return new Tuple<TreeNode, int> (null, 0);
        }
        
        var left = dfs(node.left); 
        var right = dfs(node.right);
        
        if (left.Item2 > right.Item2) {
            return new Tuple<TreeNode, int> (left.Item1, left.Item2 + 1);
        }
        else if (left.Item2 < right.Item2) {
             return new Tuple<TreeNode, int> (right.Item1, right.Item2 + 1);
        }
        else {
            return new Tuple<TreeNode, int> (node, right.Item2 + 1);
        }
    }
}

File: Solutions/865-Smallest Subtree with all the Deepest Nodes/865-Smallest Subtree with all the Deepest Nodes.js
//https://leetcode.com/problems/smallest-subtree-with-all-the-deepest-nodes/discuss/969896/JavaScript-TypeScript-Solution
//Third fastest
var subtreeWithAllDeepest = function (root) {
  if (!root) return root;

  const left = getDepth(root.left);
  const right = getDepth(root.right);

  if (left > right) {
    return subtreeWithAllDeepest(root.left);
  }

  if (left < right) {
    return subtreeWithAllDeepest(root.right);
  }

  return root;
};

const getDepth = (node) => {
  if (!node) return 0;

  const left = getDepth(node.left);
  const right = getDepth(node.right);

  return Math.max(left, right) + 1;
};

//Another
var subtreeWithAllDeepest = function(root) {   
    return dfs(root, 0)[0];        
};


function dfs(root, depth) {
    if(root == null) return [null, depth];
    
    const leftNode = dfs(root.left, depth + 1);
    const rightNode = dfs(root.right, depth + 1);
    
    if (leftNode[1] === rightNode[1]) {
        return [root, leftNode[1]];
    } else {
        return leftNode[1] < rightNode[1] ? rightNode : leftNode;
    }
};

//Second fastest
const subtreeWithAllDeepest = (root) => {
	const findLCA = (node) => {
		if (node === null) return false;
			if (node.val in hash) {
			if (maxLeaves.length === 1) LCA = node;
			return true;
		}

		var left = findLCA(node.left),
			right = findLCA(node.right);

		if (left && right) {
			LCA = node;
			return true;
		} else if (left || right) {
			return true;
		}
		return false;
	};

	const findDeepestNodes = (node, depth) => {
		if (node === null) return null;

		depth++;

		let left = findDeepestNodes(node.left, depth),
			right = findDeepestNodes(node.right, depth);

		// Is it a leaf?
		if (left === null && right === null) {
			if (depth > maxDepth) {
				maxDepth = depth;
				maxLeaves = [node];
			} else if (depth === maxDepth) {
				maxLeaves.push(node);
			}
		}

		depth--;
		return [maxDepth, maxLeaves];
	};

	var maxDepth = -1,
		maxLeaves = [];
	// Find deepest leaves or leaf
	[maxDepth, maxLeaves] = findDeepestNodes(root, -1);

	var hash = {};
	maxLeaves.forEach((leaf) => {
		hash[leaf.val] = true;
	});

	var LCA = root;
	// Find parent that has both in left and right subtree
	findLCA(root);

	return LCA;
};

//Fastest
var subtreeWithAllDeepest = function(root) {
    function _dfs(node) {
      if (!node) return new Result(null, 0);
      
      let left = _dfs(node.left);
      let right = _dfs(node.right);
      if (left.dist < right.dist) return new Result(right.node, right.dist+1);
      if (left.dist > right.dist) return new Result(left.node, left.dist+1);
      return new Result(node, left.dist+1);
    } 
    
    class Result {
      constructor(node, dist) {
        this.node = node;
        this.dist = dist;
      }
    }
    
    return _dfs(root).node;
  };